envoy/stream_info/filter_state.h

```diff
@@ -103,8 +109,17 @@ class FilterState {
 
   /**
    * @param data_name the name of the data being looked up (mutable/readonly).
-   * @return a const reference to the stored data.
-   * An exception will be thrown if the data does not exist.
+   * @return a typed pointer to the stored data or nullptr if the data does not exist or the data
+   * type does not match the expected type.
+   */
+  template <typename T> const T* getDataReadOnlyTyped(absl::string_view data_name) const {
+    const auto* generic_result = getDataReadOnlyGeneric(data_name);
+    return generic_result == nullptr ? nullptr : dynamic_cast<const T*>(generic_result);
```

Comment:
It seems odd to me that the pointer return version of this API doesn't check dynamic_cast success and then throws, but the reference version does. 2 questions:
1) Should we have both versions? Can we only have 1?
2) Should they both throw if we have both?

Reply:
Yes. We can only keep one of these method. In fact, I think the second version (no throws) is more valuable. Because in current Envoy, we never catch the exception but always use code pattern like this:
```
if(filter_state.hasData<Type>(name)) {
  filter_state.get.....;
}
```

If you think it is ok, I can try to land it in this PR. But this could lead to the PR becoming even bigger.

Reply:
Yup let's just go ahead and clean it up, thank you.

/wait